src/preprocessing/passes/real_to_int.cpp

```diff
@@ -0,0 +1,205 @@
+              if (v.isNull())
+              {
+                sum.push_back(c);
+              }
+              else
+              {
+                Node vv = realToIntInternal(v, cache, var_eq);
+                if (vv.getType().isInteger())
+                {
+                  sum.push_back(
+                      NodeManager::currentNM()->mkNode(kind::MULT, c, vv));
+                }
+                else
+                {
+                  throw TypeCheckingException(
+                      v.toExpr(),
+                      string("Cannot translate to Int: ") + v.toString());
+                }
+              }
+            }
+            Node sumt =
+                sum.empty()
+                    ? NodeManager::currentNM()->mkConst(Rational(0))
+                    : (sum.size() == 1
+                           ? sum[0]
+                           : NodeManager::currentNM()->mkNode(kind::PLUS, sum));
+            ret = NodeManager::currentNM()->mkNode(
+                ret_lit.getKind(),
+                sumt,
+                NodeManager::currentNM()->mkConst(Rational(0)));
+            if (!ret_pol)
+            {
+              ret = ret.negate();
+            }
+            Trace("real-as-int") << "Convert : " << std::endl;
+            Trace("real-as-int") << "   " << n << std::endl;
+            Trace("real-as-int") << "   " << ret << std::endl;
+          }
+          else
+          {
+            throw TypeCheckingException(
+                n.toExpr(), string("Cannot translate to Int: ") + n.toString());
+          }
+        }
+      }
+      else
+      {
+        bool childChanged = false;
+        std::vector<Node> children;
+        for (unsigned i = 0; i < n.getNumChildren(); i++)
+        {
+          Node nc = realToIntInternal(n[i], cache, var_eq);
+          childChanged = childChanged || nc != n[i];
+          children.push_back(nc);
+        }
+        if (childChanged)
+        {
+          ret = NodeManager::currentNM()->mkNode(n.getKind(), children);
+        }
+      }
+    }
+    else
+    {
+      if (n.isVar())
+      {
+        if (!n.getType().isInteger())
+        {
+          ret = NodeManager::currentNM()->mkSkolem(
+              "__realToIntInternal_var",
+              NodeManager::currentNM()->integerType(),
+              "Variable introduced in realToIntInternal pass");
+          var_eq.push_back(n.eqNode(ret));
+          TheoryModel* m = d_preprocContext->getTheoryEngine()->getModel();
+          m->addSubstitution(n, ret);
+        }
+      }
+    }
+    cache[n] = ret;
+    return ret;
+  }
+}
+
+RealToInt::RealToInt(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "real-to-int"){};
+
+PreprocessingPassResult RealToInt::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  unordered_map<Node, Node, NodeHashFunction> cache;
+  for (unsigned i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    std::vector<Node> var_eq;
```

Comment:
This seems to change the semantics of the original PP pass where `var_eq` is preserved between iterations.